### PR TITLE
Do not inherit open FDs to SSH child process

### PR DESCRIPTION
**Pull request by clue**
This somewhat obscure PR ensures that we do not inherit open file descriptors (FDs) to the SSH child process. This can cause all sorts of errors in long running applications and really is not desired here.

This is implemented by explicitly closing all superfluous FDs in the implicit `sh` child process before launching the actual `ssh` binary. PHP does not support `FD_CLOEXEC`, `O_CLOEXEC` or `SOCK_CLOEXEC`  and this appears to be the best work around I could find (yes, I should probably write a lengthy, somewhat technical blog post about this). Additionally, this PR includes a test to verify this works on all supported platforms and this could perhaps be used as a starting point for other libraries (YMMV).

See also https://github.com/reactphp/child-process/issues/51

### CODE REVIEW

File: src/SshProcessConnector.php [hunk 1]
```diff
@@ -66,7 +66,27 @@ public function connect($uri)
             return \React\Promise\reject(new \InvalidArgumentException('Invalid target URI'));
         }
 
-        $process = new Process($this->cmd . ' -W ' . \escapeshellarg($parts['host'] . ':' . $parts['port']));
+        $command = $this->cmd . ' -W ' . \escapeshellarg($parts['host'] . ':' . $parts['port']);
+
+        // try to get list of all open FDs (Linux only) or simply assume range 3-1024 (FD_SETSIZE)
+        $fds = @scandir('/proc/self/fd');
```
Comment (WyriHaximus) [inline src/SshProcessConnector.php]: FYI working on a package to do this cross platform: https://github.com/WyriHaximus/php-file-descriptors

Comment (clue) [reply src/SshProcessConnector.php]: @WyriHaximus Thank you, will definitely keep an eye on this! :+1:

**Comment by clue** [comment]
For the reference: This PR ensures that we immediately *close* all inherited file descriptors before executing the actual `ssh` child process. This means that there is still a (very short) period where these file descriptors are in fact inherited to the intermediary `sh` process and could potentially cause problems. While I have not seen any actual issues in my application because of this possible race condition, I still agree that this is something to address in a future version, most likely by taking advantage of https://github.com/reactphp/child-process/pull/65.

File: src/SshProcessConnector.php (hunk 1 above)
Comment (WyriHaximus) [reply src/SshProcessConnector.php]: @clue currently onto the daunting task of figuring it out for `Windows`...

**Pull request merged**